test/plugin/test_in_tail.rb

```diff
@@ -2638,4 +2638,278 @@ def test_lines_collected_with_no_throttling(data)
       end
     end
   end
+
+  # https://github.com/fluent/fluentd/issues/3614
+  sub_test_case "Update watchers for rotation with follow_inodes" do
+    def test_updateTW_before_refreshTW_and_detach_before_refreshTW
+      config = config_element(
+        "ROOT",
+        "",
+        {
+          "path" => "#{@tmp_dir}/tail.txt*",
+          "pos_file" => "#{@tmp_dir}/tail.pos",
+          "tag" => "t1",
+          "format" => "none",
+          "read_from_head" => "true",
+          "follow_inodes" => "true",
+          # In order to detach the old watcher quickly.
+          "rotate_wait" => "1s",
+          # In order to reproduce the same condition stably, ensure that `refresh_watchers` is not
+          # called by a timer.
+          "refresh_interval" => "1h",
+          # stat_watcher often calls `TailWatcher::on_notify` faster than creating a new log file,
+          # so disable it in order to reproduce the same condition stably.
+          "enable_stat_watcher" => "false",
+        }
+      )
+      d = create_driver(config, false)
+
+      tail_watchers = []
+      stub.proxy(d.instance).setup_watcher do |tw|
+        tail_watchers.append(tw)
+        tw
+      end
+
+      Fluent::FileWrapper.open("#{@tmp_dir}/tail.txt", "wb") {|f| f.puts "file1 log1"}
+
+      d.run(expect_records: 4, timeout: 10) do
+        # Rotate (If the timing is bad, `TailWatcher::on_notify` might be called between mv and new-file-creation)
+        Fluent::FileWrapper.open("#{@tmp_dir}/tail.txt", "ab") {|f| f.puts "file1 log2"}
+        FileUtils.move("#{@tmp_dir}/tail.txt", "#{@tmp_dir}/tail.txt" + "1")
+        Fluent::FileWrapper.open("#{@tmp_dir}/tail.txt", "wb") {|f| f.puts "file2 log1"}
+
+        # `watch_timer` calls `TailWatcher::on_notify`, and then `update_watcher` updates the TailWatcher:
+        #     TailWatcher(path: "tail.txt", inode: inode_0) => TailWatcher(path: "tail.txt", inode: inode_1)
+        # The old TailWathcer is detached here since `rotate_wait` is just `1s`.
+        sleep 3
+
+        # This reproduces the following situation:
+        #     Rotation => update_watcher => refresh_watchers
+        # This adds a new TailWatcher: TailWatcher(path: "tail.txt1", inode: inode_0)
+        d.instance.refresh_watchers
+
+        # Append to the new current log file.
+        Fluent::FileWrapper.open("#{@tmp_dir}/tail.txt", "ab") {|f| f.puts "file2 log2"}
+      end
+
+      inode_0 = tail_watchers[0].ino
+      inode_1 = tail_watchers[1].ino
+      record_values = d.events.collect { |event| event[2]["message"] }.sort
+      position_entries = []
+      Fluent::FileWrapper.open("#{@tmp_dir}/tail.pos", "r") do |f|
+        f.readlines(chomp: true).each do |line|
+          values = line.split("\t")
+          position_entries.append([values[0], values[1], values[2].to_i(16)])
+        end
+      end
+
+      assert_equal(
+        {
+          # TODO: This is BUG!! We need to fix it and replace this with the next.
```

Comment:
:thinking:

Reply:
It's not intended usage, rotated files should be excluded by `path` parameter although they are included in this test (`#{@tmp_dir}/tail.txt*`).

Reply:
It's hard to detect whether tail.txt1 is a completely new file or rotated file, since same inode with a different path is possible especially on just after deleting a file (since a newly created file is often given a same inode with the file delete just before it).
So I think we don't need to care rotated files by our code, they should be excluded by the `path` pattern.

So the test code should exclude the rotated path pattern.

Reply:
... :thinking:（That should not be.)
Sorry, probably it's my misunderstanding.
Please ignore my comment.

Reply:
Since the position is erased during [unwatch](https://github.com/fluent/fluentd/blob/b71ab139b25bfcb8cb3df87385a7f94c8f6fd220/lib/fluent/plugin/in_tail/position_file.rb#L56-L60), there is a possibility of collecting duplicate files after the rotation depending on timing.

This is for reducing memory consumption:

* #2803

However, it does not seem to work correctly with `follow_inodes`, which is implemented after this fix.

I want to fix it following this PR.